modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/consensus/ConsensusManager.scala

```diff
@@ -18,44 +18,51 @@ import org.tessellation.node.shared.infrastructure.consensus.trigger.{ConsensusT
+import org.tessellation.schema.peer.Peer
 import org.tessellation.schema.peer.Peer.toP2PContext
-import org.tessellation.schema.peer.{Peer, PeerId}
-import org.tessellation.security.hash.Hash
 import org.tessellation.security.signature.Signed
 
 import eu.timepit.refined.auto._
+import monocle.Lens
 import org.typelevel.log4cats.slf4j.Slf4jLogger
 import retry.RetryDetails
 import retry.RetryPolicies.{constantDelay, fullJitter, limitRetries}
 import retry.syntax.all._
 
-trait ConsensusManager[F[_], Key, Artifact, Context] {
+trait ConsensusManager[F[_], Key, Artifact, Context, Status, Outcome, Kind] {
   def registerForConsensus(observationKey: Key): F[Unit]
 
   def startFacilitatingAfterDownload(observationKey: Key, lastArtifact: Signed[Artifact], lastContext: Context): F[Unit]
 
-  def startFacilitatingAfterRollback(lastKey: Key, lastArtifact: Signed[Artifact], lastContext: Context): F[Unit]
+  def startFacilitatingAfterRollback(lastKey: Key, initialOutcome: Outcome): F[Unit]
 
   def withdrawFromConsensus: F[Unit]
 
   private[consensus] def facilitateOnEvent: F[Unit]
 
-  private[consensus] def checkForStateUpdate(key: Key)(resources: ConsensusResources[Artifact]): F[Unit]
+  private[consensus] def checkForStateUpdate(key: Key)(resources: ConsensusResources[Artifact, Kind]): F[Unit]
 }
 
 object ConsensusManager {
 
-  def make[F[_]: Async: Metrics: Random, Event, Key: Show: Order: Next, Artifact: Eq, Context: Eq](
+  def make[F[_]: Async: Metrics: Random, Event, Key: Show: Order: Next, Artifact: Eq, Context: Eq, Status: Eq, Outcome, Kind](
     config: ConsensusConfig,
-    consensusStorage: ConsensusStorage[F, Event, Key, Artifact, Context],
-    consensusStateCreator: ConsensusStateCreator[F, Key, Artifact, Context],
-    consensusStateUpdater: ConsensusStateUpdater[F, Key, Artifact, Context],
-    consensusStateRemover: ConsensusStateRemover[F, Key, Artifact, Context],
+    consensusStorage: ConsensusStorage[F, Event, Key, Artifact, Context, Status, Outcome, Kind],
+    consensusStateCreator: ConsensusStateCreator[F, Key, Artifact, Context, Status, Outcome, Kind],
+    consensusStateUpdater: ConsensusStateUpdater[F, Key, Artifact, Context, Status, Outcome, Kind],
+    consensusStateAdvancer: ConsensusStateAdvancer[F, Key, Artifact, Context, Status, Outcome, Kind],
+    consensusStateRemover: ConsensusStateRemover[F, Key, Event, Artifact, Context, Status, Outcome, Kind],
+    consensusOps: ConsensusOps[Status, Kind],
     nodeStorage: NodeStorage[F],
     clusterStorage: ClusterStorage[F],
-    consensusClient: ConsensusClient[F, Key, Artifact, Context],
-    selfId: PeerId
-  )(implicit S: Supervisor[F]): F[ConsensusManager[F, Key, Artifact, Context]] = {
+    consensusClient: ConsensusClient[F, Key, Outcome]
+  )(
+    implicit S: Supervisor[F],
+    _artifact: Lens[Outcome, Signed[Artifact]],
+    _context: Lens[Outcome, Context],
+    _key: Lens[Outcome, Key],
+    _trigger: Lens[Outcome, ConsensusTrigger]
```

Comment:
Lenses used to let the developer use an arbitrary `Outcome` type but still let the developer reuse functionality that relies on the particular fields/data like `Key` or `ConsensusTrigger`. I think it's better that way than a common trait.

Reply:
nice one! :clap: